Major Overhaul: VexDoc

angelusrivera opened the pull request:
An 8-hour labor of love, all ready to merge.

✅ Added Testing/Examples
✅ Smaller binary/performance changes
✅ Updated dependencies (current versions as of 09.27.25)
✅ Revamped code/added (slight) documentation
✅ Severe brain damage

angelusrivera committed:
Major Overhaul: VEXDOC

Performance, blah blah blah

angelusrivera committed:
📝 Trim verbose documentation to essential information

• Simplified module and function documentation
• Removed lengthy explanations and examples
• Kept core functionality and usage info
• Made code more readable and focused

angelusrivera committed:
Trimmed up the file_processing.rs

angelusrivera committed:
CI/CD: Worked on fixing Github Actions Pipeline

mariano-f-r commented on src/docgen/mod.rs:
```diff
@@ -65,36 +95,38 @@ impl DocGenConfig {
-                Ok(files)
+            Ok(files) => {
+                // Filter files by extension more efficiently
+                let filtered_files: Vec<PathBuf> = files
```
Good.

mariano-f-r commented on src/docgen/mod.rs:
```diff
@@ -34,25 +37,52 @@ impl DocGenConfig {
 
-        let mut error = false;
-        let mut solution = String::new();
+        let mut errors = Vec::new();
```
Since these two are in parallel, it seems to be a sign that just storing them as vectors are the wrong abstraction. Let's try to think of a better way to do this. Perhaps a `UserError` struct that the `SubcommandError::UserError` could maintain a vector of? We could pack it full of metadata.

angelusrivera replied:
You're exactly right. Looks like we have two separate vectors, errors and suggestions, that have to stay in sync. Definitely poor abstraction as well. I can just make a struct that encapsulates all of this.

mariano-f-r commented on src/docgen/mod.rs:
```diff
@@ -162,49 +238,61 @@ enum ParserState {
+    let mut comment_buffer = String::with_capacity(256); // Buffer for comment text
+    let mut code_buffer = String::with_capacity(512); // Buffer for code text
+    
+    // Pre-compute common strings to avoid allocations in hot loop
```
Good spot, much appreciated!

mariano-f-r commented on src/docgen/mod.rs:
```diff
@@ -287,15 +378,21 @@ fn create_doc(old_path: &Path, conf: &DocGenConfig) -> Result<bool, SubcommandEr
         .with_style(include_str!("styles.css"))
         .with_stylesheet(
-            "https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/styles/default.min.css",
+            "https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/styles/github-dark.min.css",
```
This documentation is supposed to be easy to print, and a dark mode seems out of place. GitHub light is fine.

angelusrivera replied:
Perfect, I just added it back

mariano-f-r commented on src/docgen/mod.rs:
```diff
@@ -287,15 +378,21 @@ fn create_doc(old_path: &Path, conf: &DocGenConfig) -> Result<bool, SubcommandEr
         .with_script_link(
             "https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/highlight.min.js",
         )
+        .with_meta([("name", "viewport"), ("content", "width=device-width, initial-scale=1.0")])
```
good catch!

mariano-f-r commented on src/docgen/tests.rs:
```diff
@@ -1,14 +1,27 @@
     let mut paths = Vec::<PathBuf>::new();
     for i in 0..item_count {
-        let item_depth: usize = thread_rng().gen_range(0..3);
+        let item_depth: usize = rng.random_range(0..3);
```
`gen_range` was deprecated, thanks for noticing!

mariano-f-r commented on src/docgen/tests.rs:
```diff
@@ -19,10 +32,11 @@ fn rand_dir_entries(path: &Path) -> Vec<PathBuf> {
-                    .map(|b| char::from(b))
+                let parent: String = (0..7)
+                    .map(|_| {
+                        let chars = b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
```
Expound on this specific design choice. `rand`'s `Alphanumeric` distribution gives us a perfectly fine way to uniformly select from alphanumeric characters.

angelusrivera replied:
Fair point, I wasn't thinking about it too much when I was running thru. New commit will revert these changes.

mariano-f-r commented on src/docgen/tests.rs:
```diff
@@ -37,15 +51,17 @@ fn rand_dir_entries(path: &Path) -> Vec<PathBuf> {
-                    .map(|b| char::from(b))
+                let parent: String = (0..7)
+                    .map(|_| {
+                        let chars = b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
```
see above.

mariano-f-r commented on src/docgen/tests.rs:
```diff
@@ -37,15 +51,17 @@ fn rand_dir_entries(path: &Path) -> Vec<PathBuf> {
-                    .map(|b| char::from(b))
+                let parent2: String = (0..7)
+                    .map(|_| {
+                        let chars = b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
```
see above above.

mariano-f-r commented on src/docgen/tests.rs:
```diff
@@ -83,19 +99,19 @@ fn random_get_all_files() -> Result<(), Box<dyn Error>> {
 #[test]
 fn reads_config() -> Result<(), Box<dyn Error>> {
-    let tmp_config = NamedTempFile::new("VexDoc.toml")?;
+    let tmp_dir = TempDir::new()?;
```
To my knowledge `NamedTempFile` creates a new temporary directory as a parent anyways. Explain this design decision?

angelusrivera replied:
Essentially `TempDir` is supposed to test simulating running the tool from a project directory where it looks for VexDoc.toml. `NamedTempFile` would just create a file in /tmp/, at least from what I've seen from my testing. Could require further testing.

mariano-f-r commented on src/docgen/tests.rs:
```diff
@@ -106,6 +122,8 @@ file_extensions = ["c", "h"]
     );
 
+    // Restore original directory
+    env::set_current_dir(original_dir)?;
```
nice catch

mariano-f-r commented on src/docgen/tests.rs:
```diff
@@ -157,8 +179,227 @@ file_extensions = []
     } = conf.unwrap_err()
     {
+        // Restore original directory
+        env::set_current_dir(original_dir)?;
```
No need to do it twice, we already have the `conf`, just do it as soon as you have that data.

mariano-f-r commented on src/docgen/tests.rs:
```diff
@@ -157,8 +179,227 @@ file_extensions = []
+}
+
+#[test]
+fn test_quiet_mode() -> Result<(), Box<dyn Error>> {
```
What you could do is refactor any pieces of code that need to provide output to take an `impl Write`. Then, you could pass in `Stdout`, but also regular string buffers to then `assert_eq` on.

mariano-f-r commented on src/main.rs:
```diff
@@ -1,13 +1,16 @@
 fn main() {
     let args: VexDocArgs = argh::from_env();
     let mut exit_code = 0;
-    // we only care about Err values because the actual results are all side effects
```
I happen to think that this is a useful comment.

angelusrivera replied:
As do I, reverting it as we speak.

mariano-f-r reviewed (changes requested):
Overall good, a few key points.

angelusrivera committed:
🔥 refactor: improve code quality and convert to print-friendly light theme

- Replace parallel error vectors with ValidationError struct
- Convert dark theme to white background for better printability
- Fix documentation annotations across source files
- Add helper function for random string generation in tests

angelusrivera committed:
fix: resolve hardcoded path in create_config function

- Change File::create_new("./VexDoc.toml") to File::create_new(&dir)
- Now properly respects the --dir parameter instead of always creating in current directory
- Fixes issue #16
